doc/design/blob-uuid-1.md

```diff
@@ -0,0 +1,51 @@
+A blob is referenced by a key generation function. the value of this key is used to index the blob in the database table of identities and on the s3 storage. `s3://dandiarchive/blob/<3digitskey>/<next3digitskey>/<key>`
+
+** database table of identities **: indexed by the key and contains columns for as many types of checksums as we want. these checksums constitute identities of the blob and should use functions that will always return the same value.
+
+**key generation function**: the current suggestion (from satra) for this function is UUID4.
+
+**DANDI ETag generation function**: this is an agreement between API server and DANDI CLI for now to ensure there is an algorithm that consistently returns the correct number of parts given a `contentSize` in bytes. this should in theory also be the AWS ETag generation function.
+
+## Upload (from a CLI)
+
+### Check process (for de-duplication):
+1. CLI `POST /blobs/digests/` <-- `{"algorithm": "dandi:dandi-etag", "value": "123etag123"}`. API:
+    1. Looks for an existing AssetBlob with the given digest(s).
+    2. If it exists, return the key (UUID).
+    3. It it doesn't exist, return 404.
+2. If 404, start upload process.
+3. if key, call `POST .../assets/` with the key to register the new asset.
+
+### Blob upload process:
+1. CLI `/uploads/initialize` <-- `{"content_size": 12345 , "digest": {"algorithm": "dandi:dandi-etag", "value": "..."}`. API:
+   1. Checks if there is already an asset blob with the given digest. If so, return 409 CONFLICT with the `blobs:key` in the header (`Location`?).
```

Comment:
@dchiquito so is it to be `Location` header field in response and thus we should remove `?` or something else?

Reply:
I implemented it to use the `Location` header, but I'm happy to leave the `?` until the CLI half is implemented in case it turns out something else would be preferable.